### ISSUE TITLE
ICE: broken MIR in Item(WithOptConstParam...) — encountered `Assign` statement with incompatible types

### ISSUE THREAD

**Issue by little-arhat** [rust-lang/rust#75597]
Seen in automatic travis build: https://travis-ci.org/github/copterust/proving-ground/builds/718201256.

(error is triggered by [this line](https://github.com/copterust/alt-stm32f30x-hal/blob/master/src/rcc.rs#L379)). 

```
$ rustc --version

rustc 1.47.0-nightly (8e21bd063 2020-08-14)

$ rustup --version

rustup 1.22.1 (b01adbbc3 2020-07-08)

$ cargo --version

cargo 1.47.0-nightly (ab32ee88d 2020-08-10)


```

### Error output

```
Compiling alt-stm32f30x-hal v0.24.0

     Running `rustc --crate-name alt_stm32f30x_hal --edition=2018 /home/travis/.cargo/registry/src/github.com-1ecc6299db9ec823/alt-stm32f30x-hal-0.24.0/src/lib.rs --error-format=json --json=diagnostic-rendered-ansi --crate-type lib --emit=dep-info,metadata,link -C opt-level=s -C panic=abort -Cembed-bitcode=no -C codegen-units=1 -C debuginfo=2 -C debug-assertions=on --cfg 'feature="default"' --cfg 'feature="device-selected"' --cfg 'feature="rt"' --cfg 'feature="stm32f303"' -C metadata=f1ce9b70b61ec746 -C extra-filename=-f1ce9b70b61ec746 --out-dir /home/travis/build/copterust/proving-ground/target/thumbv7em-none-eabihf/debug/deps --target thumbv7em-none-eabihf -L dependency=/home/travis/build/copterust/proving-ground/target/thumbv7em-none-eabihf/debug/deps -L dependency=/home/travis/build/copterust/proving-ground/target/debug/deps --extern bitrate=/home/travis/build/copterust/proving-ground/target/thumbv7em-none-eabihf/debug/deps/libbitrate-f70c4bf2280a0908.rmeta --extern bobbin_bits=/home/travis/build/copterust/proving-ground/target/thumbv7em-none-eabihf/debug/deps/libbobbin_bits-323e8db8258afea3.rmeta --extern cast=/home/travis/build/copterust/proving-ground/target/thumbv7em-none-eabihf/debug/deps/libcast-65ce1edada04de5f.rmeta --extern cortex_m=/home/travis/build/copterust/proving-ground/target/thumbv7em-none-eabihf/debug/deps/libcortex_m-b54da80a2428812b.rmeta --extern hal=/home/travis/build/copterust/proving-ground/target/thumbv7em-none-eabihf/debug/deps/libembedded_hal-c3a9158d7cbfe6b5.rmeta --extern nb=/home/travis/build/copterust/proving-ground/target/thumbv7em-none-eabihf/debug/deps/libnb-bc2b79b5f894c7f4.rmeta --extern stm32f3=/home/travis/build/copterust/proving-ground/target/thumbv7em-none-eabihf/debug/deps/libstm32f3-032fe9becdb87096.rmeta --extern void=/home/travis/build/copterust/proving-ground/target/thumbv7em-none-eabihf/debug/deps/libvoid-2579d08d7437c761.rmeta --cap-lints allow -C link-arg=-Tlink.x -L /home/travis/build/copterust/proving-ground/target/thumbv7em-none-eabihf/debug/build/cortex-m-9396aca9416fc59e/out -L /home/travis/build/copterust/proving-ground/target/thumbv7em-none-eabihf/debug/build/cortex-m-adda42040cb9fc18/out -L /home/travis/build/copterust/proving-ground/target/thumbv7em-none-eabihf/debug/build/stm32f3-ca9fdda15b6dd6d4/out -L /home/travis/build/copterust/proving-ground/target/thumbv7em-none-eabihf/debug/build/cortex-m-rt-2e52a225fbf36dae/out`

error: internal compiler error: broken MIR in Item(WithOptConstParam { did: DefId(0:502 ~ alt_stm32f30x_hal[6093]::rcc[0]::{{impl}}[4]::freeze[0]), const_param_did: None }) (end of phase Optimized) at bb66[4]:

encountered `Assign` statement with incompatible types:

left-hand side has type: u8

right-hand side has type: bool

   --> /home/travis/.cargo/registry/src/github.com-1ecc6299db9ec823/alt-stm32f30x-hal-0.24.0/src/rcc.rs:379:17

    |

379 |                 0b01

    |                 ^^^^

    |

    = note: delayed at src/librustc_mir/transform/validate.rs:140:36

thread 'rustc' panicked at 'no errors encountered even though `delay_span_bug` issued', src/librustc_errors/lib.rs:366:17

note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

**Comment by LeSeulArtichaut** [comment]
I can't reproduce this issue on latest nightly (7e6d6e5f5 2020-08-16), using the following commands:
```bash
git clone https://github.com/copterust/proving-ground.git
cd proving-ground
rustup override set nightly
rustup target add thumbv7em-none-eabihf
mkdir gcc
curl -L https://developer.arm.com/-/media/Files/downloads/gnu-rm/7-2018q2/gcc-arm-none-eabi-7-2018-q2-update-linux.tar.bz2\?revision\=bc2c96c0-14b5-4bb4-9f18-bceb4050fee7\?product\=GNU%20Arm%20Embedded%20Toolchain,64-bit,,Linux,7-2018-q2-update | tar --strip-components=1 -C gcc -xj
export PATH="$PATH:$PWD/gcc/bin"
make mem=128k nodevice
RUST_BACKTRACE=1 make mem=128k all
```

However, I can reproduce it on nightly-2020-08-14 with the same commands, so this issue appears to already be fixed. @little-arhat are you still encountering issues with the latest nightly?

**Comment by little-arhat** [comment]
@LeSeulArtichaut  indeed, this seems to be fixed already, thanks!

**Comment by LeSeulArtichaut** [comment]
Would this be worth finding a test to prevent future regressions? cc @wesleywiser

**Comment by wesleywiser** [comment]
@LeSeulArtichaut Yeah, that seems like a great idea!

**Comment by LeSeulArtichaut** [comment]
The ICE desappears in nightly-2020-08-16, so the fix should be contained in 8e21bd0633b8d970646ee6eb706c9e8acfad19af...9b88e0a8667040452a94fda8548f0f5f0f801f90.

**Comment by JulianKnodt** [comment]
Ah sorry I suspect this was fixed here by: https://github.com/rust-lang/rust/pull/75537
And this might be a replica of other issues linked in that PR. I think there are other regression tests included in that PR, so I believe this should be resolved.
